File: VB/CustomEditingFieldsSelector/TypeBasedEditorTemplateSelector.vb
Imports DevExpress.Xpf.Printing
Imports DevExpress.XtraPrinting
Imports System
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Data
Imports System.Windows.Markup

Namespace CustomEditingFieldsSelector

    Public Class TypeBasedEditorTemplateSelector
        Inherits EditingFieldTemplateSelector

        Public Property Int32Template As DataTemplate

        Public Property DateTimeTemplate As DataTemplate

        Public Property EnumTemplate As DataTemplate

        Public Overrides Function SelectTemplate(ByVal item As Object, ByVal container As DependencyObject) As DataTemplate
            Dim field = TryCast(item, EditingField)
            If field.EditValue IsNot Nothing AndAlso TypeOf field.EditValue Is [Enum] Then
                Return EnumTemplate
            ElseIf TypeOf field.EditValue Is Date Then
                Return DateTimeTemplate
            ElseIf TypeOf field.EditValue Is Integer Then
                Return Int32Template
            End If

            Return MyBase.SelectTemplate(item, container)
        End Function
    End Class

    Public Class EnumValueToEnumSourceConverter
        Inherits MarkupExtension
        Implements IValueConverter

        Public Overrides Function ProvideValue(ByVal serviceProvider As IServiceProvider) As Object
            Return Me
        End Function

        Private Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            If Not(TypeOf value Is [Enum]) Then Throw New InvalidOperationException()
            Return [Enum].GetValues(value.GetType())
        End Function

        Private Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotSupportedException()
        End Function
    End Class
End Namespace

File: CS/CustomEditingFieldsSelector/TypeBasedEditorTemplateSelector.cs
using DevExpress.Xpf.Printing;
using DevExpress.XtraPrinting;
using System;
using System.Globalization;
using System.Windows;
using System.Windows.Data;
using System.Windows.Markup;

namespace CustomEditingFieldsSelector {
    public class TypeBasedEditorTemplateSelector : EditingFieldTemplateSelector {
        public DataTemplate Int32Template { get; set; }
        public DataTemplate DateTimeTemplate { get; set; }
        public DataTemplate EnumTemplate { get; set; }


        public override DataTemplate SelectTemplate(object item, DependencyObject container) {
            var field = item as EditingField;

            if(field.EditValue != null && field.EditValue is Enum) {
                return EnumTemplate;
            } else if(field.EditValue is DateTime) {
                return DateTimeTemplate;
            } else if(field.EditValue is Int32) {
                return Int32Template;
            }

            return base.SelectTemplate(item, container);
        }
    }

    public class EnumValueToEnumSourceConverter : MarkupExtension, IValueConverter {
        public override object ProvideValue(IServiceProvider serviceProvider) {
            return this;
        }
        object IValueConverter.Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            if(!(value is Enum))
                throw new InvalidOperationException();
            return Enum.GetValues(value.GetType());
        }

        object IValueConverter.ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotSupportedException();
        }
    }
}
